src/sdi-refresh-monitor.c

```diff
@@ -15,181 +15,523 @@
-    handle_extra_params(self, dialog, extra_params);
+static void manage_refresh_inhibit(SnapdClient *source, GAsyncResult *res,
+                                   gpointer p) {
+  g_autoptr(SdiRefreshMonitor) self = p;
+
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GPtrArray) snaps =
+      snapd_client_get_snaps_finish(source, res, &error);
+
+  if (error != NULL) {
+    if (g_error_matches(error, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
+      return;
+    }
+    g_print("Error in refresh inhibit: %s\n", error->message);
+    return;
   }
+  if (snaps->len == 0)
+    return;
+  // remove snaps that are marked as "ignore"
+  g_autoptr(GSList) snaps_not_ignored = NULL;
+  for (int i = 0; i < snaps->len; i++) {
+    const gchar *name = snapd_snap_get_name(snaps->pdata[i]);
+    if (name == NULL)
+      continue;
+    g_autoptr(SdiSnap) snap_data = add_snap(self, name);
+    if (snap_data == NULL)
+      continue;
+    sdi_snap_set_inhibited(snap_data, TRUE);
+    if (sdi_snap_get_ignored(snap_data))
+      continue;
+    snaps_not_ignored =
+        g_slist_prepend(snaps_not_ignored, g_object_ref(snaps->pdata[i]));
+  }
+  if (g_slist_length(snaps_not_ignored) > 1) {
+    sdi_notify_pending_refresh_multiple(self->notify, snaps_not_ignored);
+    return;
+  }
+  if (snaps_not_ignored != NULL) {
+    // just one notice
+    sdi_notify_pending_refresh_one(self->notify, snaps_not_ignored->data);
+    return;
+  }
+}
 
-  snap_desktop_integration_complete_application_refresh_percentage(
-      self->skeleton, invocation);
+static void notice_cb(GObject *object, SnapdNotice *notice, gboolean first_run,
+                      gpointer monitor) {
+  SdiRefreshMonitor *self = SDI_REFRESH_MONITOR(monitor);
 
-  return TRUE;
+  gboolean sent_refresh_inhibit = FALSE;
+
+  g_autoptr(GHashTable) data = snapd_notice_get_last_data(notice);
+  g_autofree gchar *kind = g_strdup(g_hash_table_lookup(data, "kind"));
+
+  switch (snapd_notice_get_notice_type(notice)) {
+  case SNAPD_NOTICE_TYPE_CHANGE_UPDATE:
+    if (!g_str_equal(kind, "auto-refresh"))
+      return;
+    snapd_client_get_change_async(
+        self->client, snapd_notice_get_key(notice), NULL,
+        (GAsyncReadyCallback)manage_change_update, g_object_ref(self));
+    break;
+  case SNAPD_NOTICE_TYPE_REFRESH_INHIBIT:
+    /**
+     * in each refresh-inhibit notice, all refresh-inhibited snaps will be
+     * processed, so we should only do it once during startup, to avoid showing
+     * several notices. Instead, only one with all the pending snaps must be
+     * shown to the user.
+     */
+    if (!first_run || !sent_refresh_inhibit) {
+      sent_refresh_inhibit = TRUE;
+      snapd_client_get_snaps_async(
+          self->client, SNAPD_GET_SNAPS_FLAGS_REFRESH_INHIBITED, NULL, NULL,
+          (GAsyncReadyCallback)manage_refresh_inhibit, g_object_ref(self));
+    }
+    break;
+  case SNAPD_NOTICE_TYPE_SNAP_RUN_INHIBIT:
+    // TODO
```

Comment:
Expand comment to indicate why this is not required to be implemented now and when it would in the future.

Reply:
Done.